gtars/src/digests/mod.rs

```diff
@@ -0,0 +1,188 @@
+//! # Fast digest computations for genomic sequences
+//!
+//! This module provides functions for computing digests of strings.
+//!
+//! # Functions
+//!
+//! The following functions are available:
+//!
+//! * `sha512t24u` - Processes a given string to compute its GA4GH sha512t24 checksum.
+//! * `md5` - Processes a given string to compute its MD5 checksum.
+//! * `digest_fasta` - Processes a FASTA file to compute the digests of each sequence in the file.
+//!
+//! # Usage
+//!
+//! The `sha512t24u` function can be used to compute the GA4GH sha512t24 checksum of a string.
+//! 
+//! ```rust
+//! use gtars::digests::sha512t24u;
+//!
+//! let digest = sha512t24u("hello world")
+//! ```
+use sha2::{Digest, Sha512};
+use md5::Md5;
+use seq_io::fasta::{Reader, RefRecord, Record};
+use std::io::prelude::{Read, Write};
+use std::fs::File;
+use flate2::read::MultiGzDecoder;
+use std::io;
+
+
+/// A struct representing the digest of a given string.
+#[derive(Debug)]
+pub struct DigestResult {
+    pub id: String,
+    pub length: usize,
+    pub sha512t24u: String,
+    pub md5: String,
+}
+
+
+/// Processes a given string to compute its GA4GH sha512t24u digest.
+///
+/// # Arguments
+///
+/// * `string` - The input string to be processed.
+///
+/// # Returns
+///
+/// A string SHA-512 digest of the input string.
+pub fn sha512t24u(string: &str) -> String {
+    let mut sha512_hasher_box = Box::new(Sha512::new());
+    for s in string.as_bytes().chunks(800) {
+        sha512_hasher_box.as_mut().update(s);
+    }
+    base64_url::encode(&sha512_hasher_box.as_mut().finalize_reset()[0..24])
+}
+
+/// Process a string to compute its md5 digest
+///
+/// # Arguments
+///
+/// * `string` - The input string to be processed.
+///
+/// # Returns
+///
+/// A string MD5 digest of the input string.
+pub fn md5(string: &str) -> String {
+    let mut hasher = Md5::new();
+    for s in string.as_bytes().chunks(800) {
+        hasher.update(s);
+    }
+    let result = hasher.finalize();
+    format!("{:x}", result)
+}
+
+/// Returns a `Read` object for a given file path.
+fn get_file_reader(file_path: &str) -> Result<Box<dyn Read>, io::Error> {
+    if file_path == "-" {
+        Ok(Box::new(std::io::stdin()) as Box<dyn Read>)
+    } else if file_path.ends_with(".gz") {
+        let file = File::open(file_path)?;
+        Ok(Box::new(MultiGzDecoder::new(file)) as Box<dyn Read>)
+    } else {
+        let file = File::open(file_path)?;
+        Ok(Box::new(file) as Box<dyn Read>)
+    }
+}
```

Comment:
We have something similar in utils, should we try to use that instead? It returns a Reader though not a Read.
https://github.com/databio/gtars/blob/bfcabdc8511b0bd1674759dfc6ace847fb5157db/gtars/src/common/utils.rs#L21-L33

Reply:
Yeah I was going to mention that we have a generic utility for this I tink

Reply:
hm; I'd have to look into what are the differences between a `Read` and a `Reader` -- it it looks like yours is also returning a `Read` object....
One uses `MultiGzDecoder` and the other is `GzDecoder`

Reply:
Ok I integrated these.